pkg/signature/kms/ehsm/signer.go

```diff
@@ -0,0 +1,213 @@
+package ehsm
+
+import (
+	"context"
+	"crypto"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+// nolint:revive
+const (
+	AlgorithmEHRSA4096 = "EH_RSA_4096"
+	AlgorithmEHRSA3072 = "EH_RSA_3072"
+	AlgorithmEHRSA2048 = "EH_RSA_2048"
+	AlgorithmEHECP256   = "EH_EC_P256"
+	AlgorithmEHECP224   = "EH_EC_P224"
+	AlgorithmEHECP384   = "EH_EC_P384"
+	AlgorithmEHECP521   = "EH_EC_P521"
+	AlgorithmEHEMSM2   = "EH_SM2"
+)
+
+var ehsmSupportedAlgorithms = []string{
+	AlgorithmEHRSA4096,
+	AlgorithmEHRSA3072,
+	AlgorithmEHRSA2048,
+	AlgorithmEHECP256,
+	AlgorithmEHECP224,
+	AlgorithmEHECP384,
+	AlgorithmEHECP521,
+	AlgorithmEHEMSM2,
+}
+
+var ehsmSupportedHashFuncs = []crypto.Hash{
+	crypto.SHA224,
+	crypto.SHA256,
+	crypto.SHA384,
+	crypto.SHA512,
+	crypto.Hash(0),
+}
+
+// SignerVerifier creates and verifies digital signatures over a message using EHSM KMS service
+type SignerVerifier struct {
+	hashFunc crypto.Hash
+	client   *ehsmClient
+}
+
+// LoadSignerVerifier generates signatures using the specified key object in Ehsm and hash algorithm.
+//
+// It also can verify signatures (via a remote vall to the Ehsm instance). hashFunc should be
+// set to crypto.Hash(0) if the key referred to by referenceStr is an ED25519 signing key.
+func LoadSignerVerifier(referenceStr string, hashFunc crypto.Hash, opts ...signature.RPCOption) (*SignerVerifier, error) {
+	h := &SignerVerifier{}
+	ctx := context.Background()
+	rpcAuth := options.RPCAuth{}
+	var keyVersion string
+	for _, opt := range opts {
+		opt.ApplyRPCAuthOpts(&rpcAuth)
+		opt.ApplyContext(&ctx)
+		opt.ApplyKeyVersion(&keyVersion)
+	}
+
+	var keyVersionUint uint64
+	var err error
+	if keyVersion != "" {
+		keyVersionUint, err = strconv.ParseUint(keyVersion, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("parsing key version: %w", err)
+		}
+	}
+
+	h.client, err = newEhsmClient(rpcAuth.Address, rpcAuth.Token, rpcAuth.Path, referenceStr, keyVersionUint)
+	if err != nil {
+		return nil, err
+	}
+
+	switch hashFunc {
+	case 0, crypto.SHA224, crypto.SHA256, crypto.SHA384, crypto.SHA512:
+		h.hashFunc = hashFunc
+	default:
+		return nil, errors.New("hash function not supported by Ehsm")
+	}
+
+	return h, nil
+}
+
+// SignMessage signs the provided message using HashiCorp Ehsm KMS. If the message is provided,
```

Comment:
check and remove "HashiCorp" in the whole file

Reply:
ok